apollo-spaceport/src/server.rs

```diff
@@ -336,24 +354,24 @@ async fn extract_graph_usage(
     let mut all_entries = task_graph_usage.lock().await;
     let drained = all_entries
         .drain()
-        .filter(|(_graph, graph_usage)| !graph_usage.is_empty())
-        .collect::<Vec<(ReporterGraph, QueryUsageMap)>>();
+        .collect::<Vec<(ReporterGraph, QueryUsage)>>();
     // Release the lock ASAP so that clients can continue to add data
     drop(all_entries);
     let mut results = Vec::with_capacity(drained.len());
-    for (graph, graph_usage) in drained {
-        tracing::info!("submitting: {} records", graph_usage.len());
-        tracing::debug!("containing: {:?}", graph_usage);
+    for (graph, query_usage) in drained {
+        tracing::debug!("submitting: {} operations", query_usage.operation_count);
+        tracing::debug!("containing: {:?}", query_usage);
         match crate::Report::try_new(&graph.reference) {
```

Comment:
This appears to create a Report whose header only has the graph_ref field set.

Notably, it doesn't set the executable_schema_id field. This field should be set to the hex sha256 of the text of the supergraph schema, I believe.  I have to admit I'm not 100% sure what aspects of our system actually depend on this being sent but it sure seems like a good idea.

This raises the broader point — it doesn't look like in general you're tying "the current schema" to any of this reporting stuff. I know that you implement Uplink with polling and there's a `reload_server` thing that happens when the schema changes, but I'm not sure how much of the system is "reloaded" when that happens. Also my understanding is that the model is that the Spaceport thing can run sorta separately from the main server.  So it seems like a Spaceport could potentially get different messages about operations run against different schema versions?

In addition to just making sure that header.executable_schema_id is set correctly, the referenced_fields_by_type data structure depends on the schema.  I know I've said (and you've implemented) "this structure is super cacheable, it only depends on the operation and the schema" but ... it does depend on the schema :) It doesn't look like the messages in agents.proto talk about the schema version.

So you could have the schema `type A { y: ID } type Query { x: A }` and the operation `{x{y}}`. This operation has referenced fields `Query.x` and `A.y`. Now you update to the schema `type B { y: ID } type Query {x: B}`. The same operation now has referenced fields `Query.x` and `B.y`. So it's not kosher to combine these two into the same message, as the referenced_fields data will be wrong for one of them!

Reply:
#1012

Reply:
Just adding a little context here. We need to add `executable_schema_id`, to our payloads but we don't need to worry about cross contamination on schema reload. The entire plugin pipeline is recreated when this happens thus we can guarantee that a particular instance will only ever be dealing with one schema.

This would not be the case in any sort of multi-tenancy scenario, but that is something to think about later.

Reply:
That's great. And that includes the "spaceport"?